<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Counting of valid read pairs between pairs of restriction fragments &mdash; Diachromatic 0.7.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=c98573da"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Summarize results" href="summarize.html" />
    <link rel="prev" title="Mapping and categorization of Hi-C reads" href="mapping.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Diachromatic
          </a>
              <div class="version">
                0.7
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Program setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial with test data</a></li>
<li class="toctree-l1"><a class="reference internal" href="digest.html">In silico genome digest</a></li>
<li class="toctree-l1"><a class="reference internal" href="truncate.html">Truncation of chimeric Hi-C reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping paired-end Hi-C reads</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Counting unique valid pairs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#required-input-files">Required input files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#gopher-digest-file">GOPHER digest file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bam-file-with-unique-valid-pairs">BAM file with unique valid pairs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#running-the-count-subcommand">Running the <em>count</em> subcommand</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">Output files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#interaction-counts">Interaction counts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#read-counts-at-interacting-fragments">Read counts at interacting fragments</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="summarize.html">Summarize results</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Diachromatic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Counting of valid read pairs between pairs of restriction fragments</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/count.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="counting-of-valid-read-pairs-between-pairs-of-restriction-fragments">
<span id="rstcount"></span><h1>Counting of valid read pairs between pairs of restriction fragments<a class="headerlink" href="#counting-of-valid-read-pairs-between-pairs-of-restriction-fragments" title="Link to this heading"></a></h1>
<p>Mapped Hi-C read pairs are typically transformed into contact matrices, whereby the pairs are counted between windows of
fixed size, typically 5 kbp (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5493985/">Forcato et al., 2017</a>) provide a review of
the methodology). Diachromatic was developed for <em>capture Hi-C</em>, which achieves
a much higher resolution than Hi-C. Therefore, for Diachromatic the read counts are determined for each
restriction digest.</p>
<section id="required-input-files">
<h2>Required input files<a class="headerlink" href="#required-input-files" title="Link to this heading"></a></h2>
<section id="gopher-digest-file">
<h3>GOPHER digest file<a class="headerlink" href="#gopher-digest-file" title="Link to this heading"></a></h3>
<p>Due to the fact that the counts are determined on the restriction fragment level, the <a class="reference internal" href="digest.html#rstdigest"><span class="std std-ref">digest file</span></a> needs to be passed to <code class="docutils literal notranslate"><span class="pre">Diachromatic</span> <span class="pre">count</span></code>. If the captured viewpoints were designed with GOPHER,
this file also includes information about active and inactive restriction fragments.</p>
</section>
<section id="bam-file-with-unique-valid-pairs">
<h3>BAM file with unique valid pairs<a class="headerlink" href="#bam-file-with-unique-valid-pairs" title="Link to this heading"></a></h3>
<p>The second required input file contains the unique valid mapped read pairs in BAM format. If this file was generated using Diachromatic with the align subcommand, nothing has to be done or taken care of. If the BAM file was produced in a different way, make sure that the two reads of any given pair occur consecutively. Furthermore, make sure that duplicates were previously removed.</p>
</section>
</section>
<section id="running-the-count-subcommand">
<h2>Running the <em>count</em> subcommand<a class="headerlink" href="#running-the-count-subcommand" title="Link to this heading"></a></h2>
<p>Use the following command to run the counting step:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ java -jar Diachromatic.jar count \
    -v prefix.valid_pairs.aligned.bam \
    -d hg19_HinDIII_DigestedGenome.txt\
    -x prefix \
    -o outdir
</pre></div>
</div>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>Short option</p></td>
<td><p>Long option</p></td>
<td><p>Example</p></td>
<td><p>Required</p></td>
<td><p>Description</p></td>
<td><p>Default</p></td>
</tr>
<tr class="row-even"><td><p>-v</p></td>
<td><p>\-\-valid-pairs-bam</p></td>
<td><p>prefix.valid_pairs.aligned.bam</p></td>
<td><p>yes</p></td>
<td><p>Path to BAM file containing unique valid pairs.</p></td>
<td><p>–</p></td>
</tr>
<tr class="row-odd"><td><p>-d</p></td>
<td><p>\-\-digest-file</p></td>
<td><p>/data/GOPHER/hg38_DpnII_DigestedGenome.txt</p></td>
<td><p>yes</p></td>
<td><p>Path to the digest file produced with GOPHER.</p></td>
<td><p>–</p></td>
</tr>
<tr class="row-even"><td><p>-o</p></td>
<td><p>\-\-out-directory</p></td>
<td><p>cd4v2</p></td>
<td><p>yes</p></td>
<td><p>Directory containing the output of the align subcommand.</p></td>
<td><p>results</p></td>
</tr>
<tr class="row-odd"><td><p>-x</p></td>
<td><p>\-\-out-prefix</p></td>
<td><p>stim_rep1</p></td>
<td><p>yes</p></td>
<td><p>Prefix for all generated files in output directory.</p></td>
<td><p>prefix</p></td>
</tr>
</tbody>
</table>
</section>
<section id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Link to this heading"></a></h2>
<p>The default name of the output file with statistics is:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">prefix.count.stats.txt</span></code></p></li>
</ul>
</div></blockquote>
<section id="interaction-counts">
<h3>Interaction counts<a class="headerlink" href="#interaction-counts" title="Link to this heading"></a></h3>
<p>The interactions are written to a tab separated text file that has the following name by default:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">prefix.interaction.counts.table.tsv</span></code></p></li>
</ul>
</div></blockquote>
<p>The structure of this file is similar to that of (<a class="reference external" href="https://bioconductor.org/packages/release/bioc/vignettes/Chicago/inst/doc/Chicago.html#ibed-format-ends-with-ibed">iBED</a>) files. Each line stands for one pair of interacting fragments. Consider the following example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr7</span>    <span class="mi">42304777</span>        <span class="mi">42314850</span>        <span class="n">A</span>       <span class="n">chr7</span>    <span class="mi">152941166</span>      <span class="mi">152943990</span>      <span class="n">I</span>       <span class="mi">14</span>
<span class="n">chr7</span>    <span class="mi">42304777</span>        <span class="mi">42314850</span>        <span class="n">A</span>       <span class="n">chr7</span>    <span class="mi">38624777</span>       <span class="mi">38625305</span>       <span class="n">I</span>       <span class="mi">11</span>
</pre></div>
</div>
<p>The first three columns contain the coordinates of a restriction fragment on chromosome 7. The <code class="docutils literal notranslate"><span class="pre">A</span></code> in column 4
indicates that this fragment is defined to be active, i.e. it is part of a viewpoint that was enriched using capture technology.
The information about active states of fragments originates from the GOPHER digest file passed to Diachromatic
using the <code class="docutils literal notranslate"><span class="pre">-d</span></code> option.</p>
<p>In addition, interactions are written to a simple pairwise interaction file <a class="reference external" href="http://wiki.wubrowse.org/Long-range">format</a> for long-range interactions established by WashU:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr13</span><span class="p">:</span><span class="mi">84250549</span><span class="o">-</span><span class="mi">84256429</span>    <span class="n">chr13</span><span class="p">:</span><span class="mi">105017710</span><span class="o">-</span><span class="mi">105020949</span>       <span class="mi">1</span>
<span class="n">chr3</span><span class="p">:</span><span class="mi">74550953</span><span class="o">-</span><span class="mi">74553110</span>     <span class="n">chr3</span><span class="p">:</span><span class="mi">83489595</span><span class="o">-</span><span class="mi">83490326</span>          <span class="mi">1</span>
</pre></div>
</div>
<p>Trans and short-range (&lt;10,000) interactions are discarded.</p>
</section>
<section id="read-counts-at-interacting-fragments">
<h3>Read counts at interacting fragments<a class="headerlink" href="#read-counts-at-interacting-fragments" title="Link to this heading"></a></h3>
<p>Another file that is created contains the counts of reads at interacting fragments. By default the name of this file is:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">prefix.interacting.fragments.counts.table.tsv</span></code></p></li>
</ul>
</div></blockquote>
<p>The structure is again similar to that of BED files. Consider the following example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr7</span>    <span class="mi">42304777</span>       <span class="mi">42314850</span>        <span class="n">A</span>       <span class="mi">25</span>
<span class="n">chr7</span>    <span class="mi">152941166</span>      <span class="mi">152943990</span>       <span class="n">I</span>       <span class="mi">14</span>
<span class="n">chr7</span>    <span class="mi">38624777</span>       <span class="mi">38625305</span>        <span class="n">I</span>       <span class="mi">11</span>
</pre></div>
</div>
<p>The first three columns contain the coordinates of interacting restriction fragments. This is again followed by either an <code class="docutils literal notranslate"><span class="pre">A</span></code> or <code class="docutils literal notranslate"><span class="pre">I</span></code> in column 4,
whereby <code class="docutils literal notranslate"><span class="pre">A</span></code> means active and <code class="docutils literal notranslate"><span class="pre">I</span></code> inactive. The fifth column contains the read counts aggregated from all
interactions that end in the corresponding fragment. For better understanding, compare these counts to the two
interactions given above.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="mapping.html" class="btn btn-neutral float-left" title="Mapping and categorization of Hi-C reads" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="summarize.html" class="btn btn-neutral float-right" title="Summarize results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, Peter Hansen, Peter Robinson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>