Search.setIndex({"docnames": ["allelspec", "count", "digest", "index", "mapping", "mapping_old", "setup", "summarize", "truncate", "tutorial", "tutorial_old"], "filenames": ["allelspec.rst", "count.rst", "digest.rst", "index.rst", "mapping.rst", "mapping_old.rst", "setup.rst", "summarize.rst", "truncate.rst", "tutorial.rst", "tutorial_old.rst"], "titles": ["Allel-specific identification of single-nucleotide polymorphism from capture Hi-C data", "Counting of valid read pairs between pairs of restriction fragments", "Creating an <em>in silico</em> restriction digest map for Diachromatic using GOPHER", "Welcome to Diachromatic\u2019s documentation", "Mapping and categorization of Hi-C reads", "Categorization of read pairs", "Program setup", "Summarize results", "Truncation of chimeric reads", "Tutorial", "&lt;no title&gt;"], "terms": {"xxx": 0, "us": [0, 1, 3, 4, 6, 8, 9], "follow": [0, 1, 3, 4, 6, 7, 8, 9], "command": [0, 1, 4, 6, 7, 8, 9], "align": [0, 1, 2, 3, 7, 9], "count": [0, 3, 4, 7], "step": [0, 1, 4, 6, 8, 9], "java": [0, 1, 3, 4, 6, 7, 8, 9], "jar": [0, 1, 4, 6, 7, 8, 9], "v": [0, 1, 9], "validpairsbamfil": 0, "op": 0, "out": [0, 1, 4, 7, 8], "directori": [0, 1, 4, 6, 7, 8, 9], "prefix": [0, 1, 4, 7, 8, 9], "map": [1, 3, 6], "hi": [1, 2, 8, 9], "c": [1, 2, 7, 8, 9], "ar": [1, 2, 4, 8, 9], "typic": 1, "transform": 1, "contact": 1, "matric": 1, "wherebi": [1, 4], "window": 1, "fix": 1, "size": [1, 2, 4, 9], "5": [1, 2, 6, 8], "kbp": 1, "forcato": 1, "et": 1, "al": 1, "2017": 1, "provid": [1, 4], "review": 1, "methodologi": 1, "diachromat": [1, 4, 7, 8, 9], "wa": [1, 2, 6, 8], "develop": 1, "captur": [1, 2, 8, 9], "which": [1, 2, 4, 9], "achiev": 1, "much": 1, "higher": [1, 6], "resolut": 1, "than": [1, 2], "therefor": [1, 2, 9], "determin": 1, "each": [1, 2], "due": [1, 4], "fact": 1, "level": [1, 4], "need": [1, 2, 6, 9], "pass": [1, 2, 9], "If": [1, 2, 4, 6], "viewpoint": [1, 3], "were": 1, "design": [1, 2, 3], "thi": [1, 2, 4, 6, 7, 9], "also": [1, 2, 9], "includ": [1, 2, 9], "inform": [1, 2], "about": [1, 2, 4, 8], "activ": 1, "inact": [1, 2], "The": [1, 2, 4, 6, 7, 8, 9], "second": [1, 4, 9], "contain": [1, 2, 4, 8, 9], "format": [1, 3], "gener": [1, 2, 4, 7, 8, 9], "noth": 1, "ha": [1, 2, 3], "done": 1, "taken": 1, "care": 1, "produc": [1, 4, 8, 9], "differ": [1, 2, 4, 8], "wai": [1, 2], "make": 1, "sure": 1, "two": [1, 4, 8], "ani": 1, "given": [1, 2, 4], "occur": [1, 8], "consecut": [1, 2, 8], "furthermor": 1, "duplic": 1, "previous": 1, "remov": [1, 8], "valid_pair": [1, 4, 9], "d": [1, 2, 4, 9], "hg19_hindiii_digestedgenom": [1, 4, 9], "txt": [1, 4, 7, 8, 9], "x": [1, 4, 7, 8, 9], "o": [1, 4, 7, 8, 9], "outdir": [1, 4, 7, 8, 9], "short": [1, 4, 7, 8], "option": [1, 2, 4, 7, 8], "long": [1, 4, 7, 8], "exampl": [1, 2, 4, 7, 8], "descript": [1, 2, 4, 7, 8], "default": [1, 2, 4, 7, 8], "ye": [1, 4, 7, 8], "path": [1, 4, 6, 7, 8, 9], "data": [1, 2, 3, 4, 7, 9], "hg38_dpnii_digestedgenom": [1, 4], "cd4v2": [1, 4, 8], "result": [1, 2, 3, 4, 8, 9], "stim_rep1": [1, 4, 8], "all": [1, 2, 3, 4, 8], "name": [1, 2, 4, 8], "statist": [1, 4, 7, 8], "i": [1, 2, 3, 4, 8, 9], "stat": [1, 4, 7, 8, 9], "written": [1, 4], "tab": 1, "separ": [1, 4], "text": 1, "tabl": [1, 2, 4], "tsv": [1, 2], "structur": 1, "similar": [1, 3], "ib": 1, "line": [1, 2, 4], "stand": 1, "one": [1, 2, 4, 9], "consid": [1, 2], "chr7": 1, "42304777": 1, "42314850": 1, "A": [1, 2, 4], "152941166": 1, "152943990": 1, "14": [1, 2], "38624777": 1, "38625305": 1, "11": [1, 2, 6], "first": [1, 2, 9], "three": 1, "column": [1, 2], "coordin": 1, "chromosom": [1, 2, 4], "7": [1, 2], "4": [1, 2, 3, 4, 9], "indic": [1, 2, 4, 6, 9], "defin": 1, "e": [1, 2, 4, 8, 9], "part": 1, "enrich": [1, 2, 3], "technologi": 1, "state": 1, "origin": [1, 4, 8], "from": [1, 2, 3, 4, 8, 9], "In": [1, 2, 3, 4, 6, 8], "addit": [1, 4, 8], "simpl": 1, "pairwis": 1, "rang": 1, "establish": 1, "washu": 1, "chr13": 1, "84250549": 1, "84256429": 1, "105017710": 1, "105020949": 1, "1": [1, 2, 4, 9], "chr3": 1, "74550953": 1, "74553110": 1, "83489595": 1, "83490326": 1, "tran": [1, 4], "10": [1, 2, 4], "000": 1, "discard": [1, 4], "anoth": 1, "creat": [1, 4, 6, 9], "By": 1, "again": 1, "bed": 1, "25": 1, "either": 1, "an": [1, 4, 7, 9], "mean": 1, "fifth": 1, "aggreg": 1, "end": [1, 2, 3, 8], "correspond": [1, 2, 4, 8], "For": [1, 6, 8], "better": 1, "understand": 1, "compar": 1, "abov": [1, 4, 9], "chc": [2, 3, 9], "protocol": [2, 3], "involv": 2, "sampl": 2, "downstream": [2, 4, 9], "analysi": [2, 4, 7, 9], "assign": 2, "read": [2, 3, 9], "pair": [2, 3, 8, 9], "subcommand": [2, 3, 9], "requir": [2, 3, 4, 6, 7, 8], "list": [2, 4], "genom": [2, 3, 4, 6, 9], "chosen": 2, "enzym": [2, 8], "Such": [2, 4], "can": [2, 4, 6, 8, 9], "softwar": 2, "export": 2, "experi": 2, "you": [2, 6, 9], "just": 2, "open": 2, "project": [2, 6], "via": 2, "menu": 2, "case": [2, 4], "non": 2, "probe": [2, 9], "associ": 2, "did": 2, "still": 2, "same": [2, 4], "It": 2, "easi": 2, "do": [2, 4], "so": 2, "setup": [2, 3, 9], "new": 2, "specifi": 2, "paramet": 2, "sourc": [2, 3, 4], "section": 2, "prepar": [2, 3], "mark": 2, "order": [2, 4], "some": [2, 9], "featur": 2, "chang": 2, "where": [2, 6], "instanc": [2, 6], "python": 2, "script": [2, 4, 9], "other": [2, 3, 6], "correspomd": 2, "consist": 2, "field": [2, 4], "describ": 2, "chr1": 2, "refer": 2, "sequenc": [2, 3, 8], "2": [2, 4, 9], "digest_start_posit": 2, "18376": 2, "base": [2, 3], "start": [2, 3], "posit": 2, "3": [2, 4, 6, 8, 9], "digest_end_posit": 2, "18392": 2, "digest_numb": 2, "42": 2, "number": [2, 4], "_restriction_sit": 2, "dpnii": [2, 3], "respons": 2, "cut": [2, 8], "6": [2, 3], "mai": [2, 4], "more": 2, "length": 2, "1245": 2, "8": 2, "_gc_content": 2, "0": 2, "500": 2, "gc": 2, "content": 2, "upstream": 2, "margin": 2, "": [2, 4, 8], "250": 2, "bp": 2, "9": 2, "_repeat_cont": 2, "138": 2, "repeat": 2, "126": 2, "12": 2, "statu": 2, "t": [2, 4, 7, 9], "f": 2, "least": [2, 4], "13": 2, "_probe": 2, "program": [3, 9], "build": [3, 9], "bowtie2": [3, 4, 9], "tutori": 3, "test": [3, 6], "dataset": 3, "truncat": [3, 4, 7], "summar": 3, "silico": 3, "digest": [3, 4, 9], "perform": [3, 8, 9], "file": [3, 6, 9], "chimer": [3, 4, 9], "run": [3, 6, 9], "output": [3, 9], "independ": 3, "forward": [3, 8], "revers": [3, 8], "properli": 3, "uniqu": [3, 4], "valid": [3, 4, 8, 9], "input": 3, "applic": [3, 6], "implement": 3, "preprocess": 3, "pipelin": [3, 6, 9], "loop": 3, "work": 3, "help": [3, 6], "relat": 3, "bait": 3, "whole": 3, "exom": 3, "set": [3, 4], "commonli": 3, "repres": [3, 4], "proxim": 3, "promot": 3, "region": [3, 4], "surround": 3, "transcript": 3, "site": [3, 8, 9], "tss": 3, "gene": 3, "interest": 3, "protein": 3, "code": 3, "detect": 3, "between": [3, 4, 9], "distal": 3, "enhanc": 3, "been": 3, "most": 3, "cutter": 3, "hindiii": [3, 8, 9], "come": 4, "interact": 4, "larg": 4, "nucleotid": [4, 8], "ci": 4, "even": 4, "locat": [4, 8, 9], "r1": [4, 8], "r2": [4, 8], "have": 4, "execut": 4, "veri": 4, "sensit": 4, "cannot": 4, "stringenc": 4, "definit": 4, "multi": 4, "stringent": 4, "There": 4, "best": 4, "sam": 4, "record": 4, "tag": 4, "bowti": [4, 9], "bsu": 4, "less": 4, "score": 4, "mapq": 4, "must": 4, "30": 4, "recommend": [4, 6], "hicup": 4, "temporari": 4, "fastq": [4, 8, 9], "retain": 4, "reorder": 4, "next": 4, "iter": 4, "simultan": 4, "over": 4, "both": 4, "singl": 4, "combin": 4, "appropri": 4, "flag": 4, "reflect": 4, "rel": 4, "orient": 4, "distinguish": 4, "sever": 4, "categori": 4, "fragment": [4, 8, 9], "biolog": 4, "random": 4, "cross": 4, "ligat": [4, 8, 9], "event": 4, "b": [4, 9], "strand": 4, "un": 4, "self": 4, "inward": 4, "point": 4, "threshold": 4, "appli": 4, "remain": 4, "them": 4, "artefactu": 4, "outward": 4, "onli": 4, "howev": 4, "we": [4, 6], "observ": [4, 8], "small": 4, "proport": 4, "classifi": 4, "strang": 4, "intern": 4, "target": [4, 6, 9], "usr": [4, 9], "bin": [4, 9], "bowtie2index": [4, 6, 9], "hg19": [4, 6, 9], "q": [4, 8, 9], "truncated_r1": [4, 8, 9], "gz": [4, 8, 9], "r": [4, 8, 9], "truncated_r2": [4, 8, 9], "possibl": 4, "argument": [4, 7, 8], "fq": [4, 8], "tool": 4, "linux": [4, 6], "x86_64": 4, "index": [4, 6, 9], "hg38": 4, "gopher": [4, 9], "p": 4, "thread": 4, "num": 4, "15": 4, "j": 4, "reject": 4, "bam": [4, 9], "fals": [4, 8], "l": 4, "lower": 4, "frag": [4, 9], "limit": 4, "50": 4, "shear": 4, "u": 4, "upper": 4, "1000": 4, "ligtion": 4, "3000": 4, "k": 4, "keep": 4, "delet": 4, "cointa": 4, "rejected_pair": 4, "ul": 4, "ulsi": 4, "sl": 4, "slsi": 4, "too": 4, "tl": 4, "vp": 4, "summari": [4, 7, 8, 9], "final": 4, "plot": 4, "shown": 4, "pdf": 4, "rscript": 4, "Or": 4, "environ": 4, "obtain": 6, "github": 6, "page": 6, "download": [6, 9], "latest": 6, "releas": 6, "should": 6, "see": [6, 9], "messag": 6, "shell": 6, "To": [6, 7, 9], "your": 6, "own": 6, "clone": 6, "repositori": 6, "app": 6, "maven": 6, "git": 6, "http": 6, "com": 6, "thejacksonlaboratori": 6, "cd": 6, "mvn": 6, "packag": 6, "whether": 6, "process": [6, 9], "success": 6, "enter": 6, "reli": 6, "instal": 6, "system": 6, "debian": 6, "sudo": 6, "apt": 6, "get": [6, 9], "prebuilt": [6, 9], "human": 6, "gb": 6, "websit": 6, "after": 6, "correct": 6, "archiv": [6, 9], "comput": 6, "unpack": 6, "unzip": [6, 9], "zip": [6, 9], "call": [6, 7, 9], "substitut": 6, "actual": 6, "html": [7, 9], "avail": [7, 8, 9], "dir": 7, "null": [7, 8], "dna": 8, "loci": 8, "link": 8, "junction": [8, 9], "search": 8, "direct": 8, "therebi": 8, "sticki": 8, "fill": [8, 9], "biotinyl": 8, "blunt": 8, "copi": 8, "overhang": 8, "restrict": [8, 9], "plain": 8, "test_1": [8, 9], "test_2": [8, 9], "symbol": 8, "true": 8, "gzip": 8, "show": 9, "how": 9, "qualiti": 9, "control": 9, "befor": 9, "proceed": 9, "pleas": 9, "instruct": 9, "well": 9, "visit": 9, "ftp": 9, "server": 9, "wget": 9, "jax": 9, "org": 9, "robinp": 9, "test_dataset": 9, "Then": 9, "decompress": 9, "gunzip": 9, "raw": 9, "recogn": 9, "juction": 9, "subdirectori": 9, "alreadi": 9, "present": 9, "detail": 9, "across": 9, "categor": 9, "note": 9, "g": 9, "grch37": 9, "onc": 9, "folder": 9, "multipl": 9, "bt2": 9, "rev": 9, "without": 9, "suffix": 9, "assum": 9, "would": 9, "crch37": 9, "intend": 9}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"allel": 0, "specif": 0, "identif": 0, "singl": 0, "nucleotid": 0, "polymorph": 0, "from": [0, 6], "captur": [0, 3], "hi": [0, 3, 4], "c": [0, 3, 4], "data": 0, "background": 0, "inform": 0, "run": [0, 1, 4, 7, 8], "diachromat": [0, 2, 3, 6], "": [0, 3], "truncat": [0, 8, 9], "subcommand": [0, 1, 4, 7, 8], "output": [0, 1, 4, 7, 8], "file": [0, 1, 2, 4, 7, 8], "count": [1, 9], "valid": 1, "read": [1, 4, 5, 8], "pair": [1, 4, 5], "between": 1, "restrict": [1, 2], "fragment": 1, "requir": 1, "input": 1, "gopher": [1, 2, 3], "digest": [1, 2], "bam": 1, "uniqu": 1, "interact": [1, 3], "creat": 2, "an": 2, "silico": 2, "map": [2, 4, 9], "us": 2, "perform": 2, "format": 2, "welcom": 3, "document": 3, "content": 3, "differenti": 3, "analysi": 3, "chromatin": 3, "gener": 3, "Of": 3, "probe": 3, "expriment": 3, "high": 3, "resolut": 3, "categor": [4, 5], "independ": 4, "forward": 4, "revers": 4, "end": 4, "properli": 4, "align": 4, "program": 6, "setup": 6, "build": 6, "sourc": 6, "prepar": 6, "bowtie2": 6, "summar": [7, 9], "result": 7, "chimer": 8, "tutori": 9, "test": 9, "dataset": 9}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 60}, "alltitles": {"Allel-specific identification of single-nucleotide polymorphism from capture Hi-C data": [[0, "allel-specific-identification-of-single-nucleotide-polymorphism-from-capture-hi-c-data"]], "Background information": [[0, "background-information"]], "Running Diachromatic\u2019s truncation subcommand": [[0, "running-diachromatic-s-truncation-subcommand"]], "Output files": [[0, "output-files"], [1, "output-files"], [4, "output-files"], [7, "output-files"], [8, "output-files"]], "Counting of valid read pairs between pairs of restriction fragments": [[1, "counting-of-valid-read-pairs-between-pairs-of-restriction-fragments"]], "Required input files": [[1, "required-input-files"]], "GOPHER digest file": [[1, "gopher-digest-file"]], "BAM file with unique valid pairs": [[1, "bam-file-with-unique-valid-pairs"]], "Running the count subcommand": [[1, "running-the-count-subcommand"]], "Interaction counts": [[1, "interaction-counts"]], "Read counts at interacting fragments": [[1, "read-counts-at-interacting-fragments"]], "Creating an in silico restriction digest map for Diachromatic using GOPHER": [[2, "creating-an-in-silico-restriction-digest-map-for-diachromatic-using-gopher"]], "Performing the in silico digest using GOPHER": [[2, "performing-the-in-silico-digest-using-gopher"]], "Format of the digest file": [[2, "format-of-the-digest-file"]], "Welcome to Diachromatic\u2019s documentation": [[3, "welcome-to-diachromatic-s-documentation"]], "Contents:": [[3, null]], "Differential Analysis of Chromatin Interactions by Capture (Diachromatic)": [[3, "differential-analysis-of-chromatin-interactions-by-capture-diachromatic"]], "Generator Of Probes for capture Hi-C Expriments at high Resolution (GOPHER)": [[3, "generator-of-probes-for-capture-hi-c-expriments-at-high-resolution-gopher"]], "Mapping and categorization of Hi-C reads": [[4, "mapping-and-categorization-of-hi-c-reads"]], "Independent mapping of forward and reverse paired-end reads": [[4, "independent-mapping-of-forward-and-reverse-paired-end-reads"]], "Pairing of properly mapped read pairs": [[4, "pairing-of-properly-mapped-read-pairs"]], "Categorization of read pairs": [[4, "categorization-of-read-pairs"], [5, "categorization-of-read-pairs"]], "Running the align subcommand": [[4, "running-the-align-subcommand"]], "Program setup": [[6, "program-setup"]], "Building Diachromatic from source": [[6, "building-diachromatic-from-source"]], "Preparation for bowtie2": [[6, "preparation-for-bowtie2"]], "Summarize results": [[7, "summarize-results"]], "Running the summarize subcommand": [[7, "running-the-summarize-subcommand"]], "Truncation of chimeric reads": [[8, "truncation-of-chimeric-reads"]], "Running the truncate subcommand": [[8, "running-the-truncate-subcommand"]], "Tutorial": [[9, "tutorial"]], "Test dataset": [[9, "test-dataset"]], "Truncation": [[9, "truncation"]], "Mapping": [[9, "mapping"]], "Counting": [[9, "counting"]], "Summarize": [[9, "summarize"]]}, "indexentries": {}})