<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial &mdash; Diachromatic 0.7.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=c98573da"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Creating an in silico restriction digest map for Diachromatic using GOPHER" href="digest.html" />
    <link rel="prev" title="Program setup" href="setup.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Diachromatic
          </a>
              <div class="version">
                0.7
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Program setup</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial with test data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#test-dataset">Test dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#truncation">Truncation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mapping">Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="#counting">Counting</a></li>
<li class="toctree-l2"><a class="reference internal" href="#summarize">Summarize</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="digest.html">In silico genome digest</a></li>
<li class="toctree-l1"><a class="reference internal" href="truncate.html">Truncation of chimeric Hi-C reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping paired-end Hi-C reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Counting unique valid pairs</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize.html">Summarize results</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Diachromatic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading"></a></h1>
<p>This tutorial shows how to use Diachromatic for processing and quality control of Capture Hi-C reads. Before proceeding with the tutorial, please follow the program setup instructions to build Diachromatic and get bowtie2 as well as the hg19 prebuilt index.</p>
<section id="test-dataset">
<h2>Test dataset<a class="headerlink" href="#test-dataset" title="Link to this heading"></a></h2>
<p>To get the data, visit this <a class="reference external" href="ftp://ftp.jax.org/robinp/Diachromatic/test_dataset/">ftp server</a> or use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">jax</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">robinp</span><span class="o">/</span><span class="n">Diachromatic</span><span class="o">/</span><span class="n">test_dataset</span><span class="o">/</span><span class="n">test_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">jax</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">robinp</span><span class="o">/</span><span class="n">Diachromatic</span><span class="o">/</span><span class="n">test_dataset</span><span class="o">/</span><span class="n">test_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">jax</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">robinp</span><span class="o">/</span><span class="n">Diachromatic</span><span class="o">/</span><span class="n">test_dataset</span><span class="o">/</span><span class="n">hg19_HinDIII_DigestedGenome</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Then decompress the digest file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gunzip</span> <span class="n">hg19_HinDIII_DigestedGenome</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="truncation">
<h2>Truncation<a class="headerlink" href="#truncation" title="Link to this heading"></a></h2>
<p>The first step of processing raw FASTQ files with Diachromatic is to recognize and truncate reads with filled-in ligation juctions, which indicate reads that include the junction of the chimeric CHC fragment. This is performed with the truncate subcommand:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ java -jar Diachromatic.jar truncate \
    -q test_1.fastq \
    -r test_2.fastq \
    -e HinDIII \
    -x prefix \
    -o outdir
</pre></div>
</div>
<p>The command will create files called <code class="docutils literal notranslate"><span class="pre">&lt;prefix&gt;.truncated_R1.fastq.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">&lt;prefix&gt;.truncated_R2.fastq.gz</span></code> in a subdirectory called <code class="docutils literal notranslate"><span class="pre">outdir</span></code> (which will be created if not already present).</p>
<p>See <a class="reference internal" href="truncate.html#rsttruncate"><span class="std std-ref">Truncation of chimeric reads</span></a> for details.</p>
</section>
<section id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Link to this heading"></a></h2>
<p>The second step of the pipeline is to map the truncated read pairs to the target genome. You also need a file that shows the locations of restriction digests across the genome. This file is included in the test dataset. You can use GOPHER to create probes and the digest file. Diachromatic uses bowtie2 to perform the mapping, and then creates a BAM file containing the valid read pairs.</p>
<p>Use the <a class="reference internal" href="mapping.html#rstmapping"><span class="std std-ref">Mapping and categorization of Hi-C reads</span></a> command to run the alignment step:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ java -jar Diachromatic.jar align \
    -b /usr/bin/bowtie2 \
    -i /path/to/bowtie2index/hg19 \
    -q prefix.truncated_R1.fastq.gz \
    -r prefix.truncated_R2.fastq.gz \
    -d hg19_HinDIII_DigestedGenome.txt \
    -x prefix \
    -o outdir
</pre></div>
</div>
<p>Note that the bowtie2 index can be downloaded from the
<a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie 2 site</a>.  The indices are available as zip archives,
e.g., <code class="docutils literal notranslate"><span class="pre">GRCh37.zip</span></code>. Once you unzip it, the resulting folder will contain multiple files (GRCh37.1.bt2, GRCh37.3.bt2, GRCh37.rev.1.bt2
GRCh37.2.bt2, GRCh37.4.bt2, GRCh37.rev.2.bt2). You need to pass the path to one of these files without the file suffix.
Assuming the directory is located at <code class="docutils literal notranslate"><span class="pre">/some/path/GRCh37</span></code>, you would therefore pass <code class="docutils literal notranslate"><span class="pre">-i</span> <span class="pre">/some/path/GRCh37/CRCh37</span></code>.</p>
<p>The diachromatic <code class="docutils literal notranslate"><span class="pre">align</span></code> command above will create a file called <code class="docutils literal notranslate"><span class="pre">&lt;prefix&gt;.valid_pairs.aligned.bam</span></code> in the <code class="docutils literal notranslate"><span class="pre">outdir</span></code> subdirectory.
See <a class="reference internal" href="mapping.html#rstmapping"><span class="std std-ref">Mapping and categorization of Hi-C reads</span></a> for details.</p>
</section>
<section id="counting">
<h2>Counting<a class="headerlink" href="#counting" title="Link to this heading"></a></h2>
<p>Use the following command to run the counting step:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ java -jar Diachromatic.jar count \
    -v prefix.valid_pairs.aligned.bam \
    -d hg19_HinDIII_DigestedGenome.txt \
    -x prefix \
    -o outdir
</pre></div>
</div>
<p>The <a class="reference internal" href="count.html#rstcount"><span class="std std-ref">Counting of valid read pairs between pairs of restriction fragments</span></a>  command produces output files intended for downstream analysis, including <code class="docutils literal notranslate"><span class="pre">&lt;prefix&gt;.count.stats.txt</span></code>
and <code class="docutils literal notranslate"><span class="pre">&lt;prefix&gt;.frag.sizes.counts.script.R</span></code>. See <a class="reference internal" href="count.html#rstcount"><span class="std std-ref">Counting of valid read pairs between pairs of restriction fragments</span></a> for details.</p>
</section>
<section id="summarize">
<h2>Summarize<a class="headerlink" href="#summarize" title="Link to this heading"></a></h2>
<p>To run the summarize command with the truncate data, use the following command.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ java -jar Diachromatic.jar summarize \
    -t outdir/prefix.truncation.stats.txt \
    -a outdir/prefix.align.stats.txt \
    -c outdir/prefix.count.stats.txt \
    -x prefix \
    -o outdir
</pre></div>
</div>
<p>The <a class="reference internal" href="summarize.html#rstsummarize"><span class="std std-ref">Summarize results</span></a> command generates an HTML file called <code class="docutils literal notranslate"><span class="pre">outdir/prefix.summary.stats.html</span></code>.</p>
<p>The summary results file for the test dataset can also be downloaded from the <a class="reference external" href="ftp://ftp.jax.org/robinp/Diachromatic/test_dataset/">ftp server</a> or use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">jax</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">robinp</span><span class="o">/</span><span class="n">Diachromatic</span><span class="o">/</span><span class="n">test_dataset</span><span class="o">/</span><span class="n">test_dataset</span><span class="o">.</span><span class="n">summary</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="setup.html" class="btn btn-neutral float-left" title="Program setup" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="digest.html" class="btn btn-neutral float-right" title="Creating an in silico restriction digest map for Diachromatic using GOPHER" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, Peter Hansen, Peter Robinson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>